diesel/src/deserialize.rs

```diff
@@ -24,6 +24,15 @@ pub type Result<T> = result::Result<T, Box<dyn Error + Send + Sync>>;
 ///
 /// This trait can be [derived](derive@Queryable)
 ///
+/// ## How to resolve compiler errors while loading data from the database
+///
+/// In case you getting uncomprehensable compiler errors while loading data
+/// from the database you might want to consider using
+/// [`#[derive(Selectable)]`](derive@crate::prelude::Selectable) +
+/// `#[diesel(check_for_backend(YourBackendType))]` to check for mismatching fields at compile
+/// time. This drastically improves the quality of the generated error messages by pointing
+/// to concrete mismatches at field level.
```

Comment:
I think, you also should describe downsides of this approach, otherwise it is unclear, why you need to manually enable great compiler messages. Also, it is unclear why the user should specify backend, because the struct is the same for all backends and isn't depend on them.

It would also be useful, if you include the example of the generated code that implements the check, maybe under the implementation note heading.

Reply:
For the downsides: There are no other downsides that this is backend specific, which should be obvious from the way you enable this.

For why not enabling that by default: 
* It would be a breaking change
* Even if it wouldn't be a breaking change, there is no way to make this work without knowing the concrete backend (which sometimes might come from a third party crate).

For the generated code: That's really uninteresting as it's just another (well placed) trait bound.

I will update the documentation to clarify why the concrete backend is required. That's because deserialization is generic over backends and rustc fails to see through this for the generic case.

Reply:
At least example of generated code here in comment could be used if someone would take its time to think about possible solutions of mentioned problems -- just to understand how exactly the solution worked right now.

Reply:
The following usage
```rust

#[derive(Selectable, Queryable)]
#[diesel(check_for_backend(diesel::pg::Pg))]
struct User {
    id: String,
    name: i32,
}

```

generates the following rust code

```rust
impl<__DB: diesel::backend::Backend> Selectable<__DB> for User {
    type SelectExpression = (users::id, users::name);
    fn construct_selection() -> Self::SelectExpression {
        (users::id, users::name)
    }
}
fn _check_field_compatibility()
where
    String: diesel::deserialize::FromSqlRow<diesel::dsl::SqlTypeOf<users::id>, diesel::pg::Pg>,
    i32: diesel::deserialize::FromSqlRow<diesel::dsl::SqlTypeOf<users::name>, diesel::pg::Pg>,
{
}

```

Without the usage of `#[diesel(check_for_backend()]` the function is not generated at all.

Reply:
Ok, it seems that the problem with the generic check is also lied in the way, that types in `schema` module is backend dependent, right? I tried to illustrait the diesel approach of mapping SQL->Rust types [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=102cf27b3d9233221a589e35594bfb9f) and it seems so.

Reply:
No, that's not the case. [This playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3b94944b4f1862d59aeafb05cccc1f9d) is closer to what diesel does. Notability the sql types can be shared between different backends (although there are also backend dependent types there). This makes the `schema.rs` module backend independent, as long as you only use types that are supported by all backends.

Reply:
Actually, you show the same thing. I'd just try to show the generic case where backend type markers are different. Of course, in reality many of them are shared.

Reply:
While that's technically true, it's an important concept that they are mostly shared in practice.